### PR TITLE
Update AZC00002 to match method pair with required requestcontext and optional convenience method

### LINKED ISSUE #6145: Update AZC0002 to match latest protocol method signature
Request comes from https://github.com/Azure/autorest.csharp/issues/3330. We need to update the analyzer.

### PR DESCRIPTION
**Pull request by pshao25**
Fix #6145

### CODE REVIEW

**Comment by weshaggard** [comment]
@JoshLove-msft would you be the right person to review this?

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs [hunk 1]
```diff
@@ -25,8 +25,17 @@ private static void CheckClientMethod(ISymbolAnalysisContext context, IMethodSym
         {
             static bool SupportsCancellationsParameter(IParameterSymbol parameterSymbol)
             {
-                return (parameterSymbol.Name == "cancellationToken" && parameterSymbol.Type.Name == "CancellationToken") ||
-                       (parameterSymbol.Name == "context" && parameterSymbol.Type.Name == "RequestContext");
+                return IsCancellationToken(parameterSymbol) || IsRequestContext(parameterSymbol);
```
Comment (m-nash) [inline src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs:28]: I am not sure we exactly want a blind OR here.  If there exists a `RequestContent` then for sure it should only have a `RequestContext` and never a `CancellationToken`.

If the response is ever `Response<T>` then we shoud only have `CancellationToken` and never `RequestContext`

Not sure if there are any other fine grained checks we can do here to make this more accurate and remove edge cases.

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0017Tests.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+using System.Threading.Tasks;
+using Xunit;
+using Verifier = Azure.ClientSdk.Analyzers.Tests.AzureAnalyzerVerifier<Azure.ClientSdk.Analyzers.ClientMethodsAnalyzer>;
+
+namespace Azure.ClientSdk.Analyzers.Tests
+{
+    public class AZC0017Tests
+    {
+        [Fact]
+        public async Task AZC0017ProducedForMethodsWithNonOptionalCancellationToken()
```
[the remark this answers is not in the document]
Comment (pshao25) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0017Tests.cs:13]: No.

AZC0002: your operations should have last parameter `CancellationToken` or `RequestContext`.
AZC0017: fine-grained rule for `CancellationToken`, your `CancellationToken` should optional, and your operation should not have `RequestContent` as parameter
AZC0018: fine-grained rule for `RequestContext`, your return type should not be a `Response<T>`.

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0018Tests.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+using Azure;
+using Azure.Core;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace RandomNamespace
+{
+    public class SomeClient
+    {
+        public virtual Task<Response> GetAsync(string s, RequestContext context = null)
```
[the remark this answers is not in the document]
Comment (pshao25) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0018Tests.cs:25]: A protocol method should return `Response`, so something like `Response<string>` should fail.

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0002Tests.cs [hunk 4]
```diff
@@ -10,7 +10,7 @@ namespace Azure.ClientSdk.Analyzers.Tests
     public class AZC0002Tests
     {
         [Fact]
-        public async Task AZC0002ProducedForMethodsWithoutCancellationToken()
```
[the remark this answers is not in the document]
Comment (pshao25) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0002Tests.cs:13]: Yes, AZC0002NotProducedForMethodsWithCancellationToken and AZC0002NotProducedForMethodsWithRequestContext

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0017Tests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (m-nash) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0017Tests.cs:13]: The original intent of AZC0002 was to validate that the last param was CancellationToken AND that it was optional which is why [this test](https://github.com/Azure/azure-sdk-tools/pull/6156/files#diff-cc0085c0e1d86252ef54bd03e81ab00485b0b68cf5236f33853bf2b325c190f5L38) existed.

I think we should keep that definition only adding the allowance for RequestContext as well.

AZC0017 should only validate that if the last param is a CancellationToken, then the method MUST NOT contain a parameter of type RequestContent

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0018Tests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (m-nash) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0018Tests.cs:25]: For return type yes, I was referring to the body param.

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0017Tests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (m-nash) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0017Tests.cs:13]: public Response<Foo> Method(string name, CancellationToken cancellationToken = default) //PASS
public Response<Foo> Method(string name, CancellationToken myToken = default) //AZC0002
public Response<Foo> Method(string name, CancellationToken cancellationToken) //AZC0002
public Response<Foo> Method(string name, RequestContent content, CancellationToken cancellationToken = default) //AZC0017
public Response<Foo> Method(string name) //AZC0002
public Response Method(string name) //AZC0002
public Response Method(string name, RequestContext context = default) //PASS
public Response Method(string name, RequestContext myContext = default) //AZC0002
public Response Method(string name, RequestContext context) //PASS
public Response Method(string name, RequestContent content, RequestContext context) //AZC0002

public Response<Foo> Method(string name, RequestContent content, RequestContext context = default) //AZC0018

//future issue
public Response Method(string name, Foo content, RequestContext context = default) //AZC0018 or AZC0019

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0002Tests.cs [hunk 5]
```diff
@@ -160,91 +152,103 @@ namespace RandomNamespace
...
             await Verifier.CreateAnalyzer(code)
                 .WithDisabledDiagnostics("AZC0015")
                 .RunAsync();
-        }
-
+        }
+
         [Fact]
-        public async Task AZC0002ProducedWhenCancellationTokenOverloadsDontMatch()
+        public async Task AZC0002NotProducedForMethodsWithRequestContentAndCancellationToken()
```
Comment (m-nash) [inline src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0002Tests.cs:197]: Is this method named incorrectly shouldn't it be `AZC0002NotProducedForMethodsWithRequestContextAndCancellationToken

**Review by m-nash — APPROVED** [review]
approved with one name typo
